### PR TITLE
ccl: introduce ccl tree and pure oss target

### PR DESCRIPTION
**Pull request by dt**
---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13013)

### CODE REVIEW

**Comment by bdarnell** [comment]
What's "DNE"?

---

Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[Makefile, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawPR0HKzUBPJkbmFeL:-KawPR0HKzUBPJkbmFeM:b-d2jvvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/Makefile#L42)):*
> ```Makefile
> STARTFLAGS   := -s type=mem,size=1GiB --alsologtostderr
> BUILDMODE    := install
> BUILDTARGET	 := .
> ```

This line contains a tab while the surrounding lines use spaces.

---

*[pkg/build/info.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawPgwV_C5NfToYfdqQ:-KawPgwV_C5NfToYfdqR:b2ly3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/pkg/build/info.go#L49)):*
> ```Go
> 	platform    = fmt.Sprintf("%s %s", runtime.GOOS, runtime.GOARCH)
> 	// Distribution is changed by the CCL init-time hook in non-APL builds.
> 	Distribution = "Apache2"
> ```

Should this be "OSS" instead of "Apache2" to match the build target and executable name? I don't feel strongly about how we name and label the builds but would prefer consistency.

---

*[pkg/ccl/buildccl/buildinfo.go, line 8 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawPmsdauAmB8hVDWac:-KawPmsdauAmB8hVDWad:bnke80x) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/pkg/ccl/buildccl/buildinfo.go#L8)):*
> ```Go
> 
> func init() {
> 	build.Distribution = "ccl"
> ```

This should probably be all caps for consistency with the capitalization of "Apache2"

---

*[pkg/cmd/cockroach-oss/main.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawOb3EBi86-tCfbanQ:-KawOb3EBi86-tCfbanR:bk2bgfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/pkg/cmd/cockroach-oss/main.go#L28)):*
> ```Go
> )
> 
> func main() {
> ```

A separate `main()` seems prone to rot. Either move the body of this function into the `cli` package so that the duplicated `main()` function is just a trivial one-liner or do everything from the top-level `main` package with the import of `ccl` controlled by a build tag. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-:-KawNDaIxocYwZ4iLlil:b7mdpd)*

**Comment by dt** [comment]
Whoops, was supposed to be DNM / do not merge (until the lic lands and I can rebase it).

---

Review status: 7 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[Makefile, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawPR0HKzUBPJkbmFeL:-KawRAoRCMHkt1DMHNyu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/Makefile#L42)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This line contains a tab while the surrounding lines use spaces.
</blockquote></details>

Done.

---

*[pkg/build/info.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawPgwV_C5NfToYfdqQ:-KawRBqPYxfh5o91GCMX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/pkg/build/info.go#L49)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should this be "OSS" instead of "Apache2" to match the build target and executable name? I don't feel strongly about how we name and label the builds but would prefer consistency.
</blockquote></details>

Done.

---

*[pkg/ccl/buildccl/buildinfo.go, line 8 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawPmsdauAmB8hVDWac:-KawUrsHVaxbpbblnxN-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/pkg/ccl/buildccl/buildinfo.go#L8)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This should probably be all caps for consistency with the capitalization of "Apache2"
</blockquote></details>

Done.

---

*[pkg/cmd/cockroach-oss/main.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawOb3EBi86-tCfbanQ:-KawUFrHL5mLTPTQKtPF:b-r7ewi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/pkg/cmd/cockroach-oss/main.go#L28)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

A separate `main()` seems prone to rot. Either move the body of this function into the `cli` package so that the duplicated `main()` function is just a trivial one-liner or do everything from the top-level `main` package with the import of `ccl` controlled by a build tag. 
</blockquote></details>

Done.

If possibly, I'd rather avoid the build tags. Also, as far as rot, we do plan to follow up with CIs to cover both `main`s.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-:-KawQg9pilMKkBtMYcmO:bcqy6qm)*

File: pkg/ccl/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Cockroach Community License (ccl) Functionality
+This tree is intended to house the bulk of our non-Apache2 licensed packages.
+By convention, all packages under this tree have the suffix `ccl`.
+
+Grouping ccl packages into one tree and clearly labeling all CCL packages
+with a recognizable suffix will hopefully make it easier to identify and prevent
+introducing any accidental dependencies on ccl from apache2 packages.
```
Comment (mjibson) [inline pkg/ccl/README.md]: Apache2 should be capitalized here since it is earlier.

File: pkg/ccl/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+
```
Comment (mjibson) [inline pkg/ccl/README.md]: Remove empty newlines.

File: pkg/ccl/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Cockroach Community License (ccl) Functionality
+This tree is intended to house the bulk of our non-Apache2 licensed packages.
+By convention, all packages under this tree have the suffix `ccl`.
+
+Grouping ccl packages into one tree and clearly labeling all CCL packages
```
Comment (mjibson) [inline pkg/ccl/README.md]: ccl has two different capitalizations

**Review by madelynnblue — COMMENTED** [review]
I'm ambivalent about build tags or a separate main.go. Otherwise LGTM.

**Comment by dt** [comment]
Review status: 6 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/ccl/README.md, line 5 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-Kat6g0xcpVN8nXllL5N-r2-5:-Kawzh2_hoDY5ykA3nhm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e2e34dbbd71125101491a136e3bc52e0b016ba/pkg/ccl/README.md#L5)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

ccl has two different capitalizations
</blockquote></details>

Done.

---

*[pkg/ccl/README.md, line 7 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-Kat6g0xcpVN8nXllL5N-r2-7:-Kawzg_EyrGL5WdBnlbJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e2e34dbbd71125101491a136e3bc52e0b016ba/pkg/ccl/README.md#L7)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Apache2 should be capitalized here since it is earlier.
</blockquote></details>

Done.

---

*[pkg/ccl/README.md, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-Kat6g0xcpVN8nXllL5N-r2-9:-KawzgLx-uYR2UNJOqtU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e2e34dbbd71125101491a136e3bc52e0b016ba/pkg/ccl/README.md#L9)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Remove empty newlines.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013)*

**Comment by danhhz** [comment]
:lgtm:

---

Review status: 6 of 15 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[main.go, line 21 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-Kax-exQ1RN3j74lhjgT:-Kax-exQ1RN3j74lhjgU:b4nfqyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f556bab8b8462348d3b28b71be2b3ca3995de1/main.go#L21)):*
> ```Go
> // The ccl hook import below means building this will produce CCL'ed binaries.
> // This file itself reamins Apache2 though to preserve the organization of ccl
> // code under the /pkg/ccl substree, but is unused for pure FLOSS builds.
> ```

subtree

---

*[pkg/ccl/ccl_init.go, line 11 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-Kax-nfjhW_9HE7zXDZE:-Kax-nfjhW_9HE7zXDZF:bi0esp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f556bab8b8462348d3b28b71be2b3ca3995de1/pkg/ccl/ccl_init.go#L11)):*
> ```Go
> package ccl
> 
> import (
> ```

comment explaining what this is, plz

---

*[pkg/cmd/cockroach-oss/main.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-KawOb3EBi86-tCfbanQ:-Kax04-yAnStJoNGG8Gr:b-f57hxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ce13f65b9efa064ef3e8c7704b549b4edd49fe/pkg/cmd/cockroach-oss/main.go#L28)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Done.

If possibly, I'd rather avoid the build tags. Also, as far as rot, we do plan to follow up with CIs to cover both `main`s.
</blockquote></details>

Agreed on the build tags. What you did seems great

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-:-Kax0FTpat2dn37LtYXV:bnfp4nl)*

**Comment by dt** [comment]
Review status: 6 of 15 files reviewed at latest revision, 9 unresolved discussions.

---

*[main.go, line 21 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-Kax-exQ1RN3j74lhjgT:-Kax7EGIT4dHgZcsOXK1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f556bab8b8462348d3b28b71be2b3ca3995de1/main.go#L21)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

subtree
</blockquote></details>

Done.

---

*[pkg/ccl/ccl_init.go, line 11 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-Kax-nfjhW_9HE7zXDZE:-Kax7Fnk1btMsL2pGu_e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f556bab8b8462348d3b28b71be2b3ca3995de1/pkg/ccl/ccl_init.go#L11)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

comment explaining what this is, plz
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013)*

**Comment by dt** [comment]
... and rebased on top of merged #13007.

**Comment by bdarnell** [comment]
:lgtm:

---

Reviewed 6 of 8 files at r2, 1 of 1 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13013#-:-KaxwhJ5eOIahNwoRXLn:bnfp4nl)*

**Pull request merged**